tests/scripts/faketime-all.sh

```diff
@@ -0,0 +1,1082 @@
+#### Initialization and command line parsing
+################################################################
+
+# Abort on errors (even on the left-hand side of a pipe).
+# Treat uninitialised variables as errors.
+set -e -o pipefail -u
+
+# Enable ksh/bash extended file matching patterns
+shopt -s extglob
+
+pre_check_environment () {
+    if [ -d library -a -d include -a -d tests ]; then :; else
+        echo "Must be run from mbed TLS root" >&2
+        exit 1
+    fi
+}
+
+pre_initialize_variables () {
+    CONFIG_H='include/mbedtls/mbedtls_config.h'
+    CRYPTO_CONFIG_H='include/psa/crypto_config.h'
+    CONFIG_TEST_DRIVER_H='tests/include/test/drivers/config_test_driver.h'
+
+    # Files that are clobbered by some jobs will be backed up. Use a different
+    # suffix from auxiliary scripts so that all.sh and auxiliary scripts can
+    # independently decide when to remove the backup file.
+    backup_suffix='.all.bak'
+    # Files clobbered by config.py
+    files_to_back_up="$CONFIG_H $CRYPTO_CONFIG_H $CONFIG_TEST_DRIVER_H"
+    # Files clobbered by in-tree cmake
+    files_to_back_up="$files_to_back_up Makefile library/Makefile programs/Makefile tests/Makefile programs/fuzz/Makefile"
+
+    append_outcome=0
+    MEMORY=0
+    FORCE=0
+    QUIET=0
+    KEEP_GOING=0
+
+    # Seed value used with the --release-test option.
+    #
+    # See also RELEASE_SEED in basic-build-test.sh. Debugging is easier if
+    # both values are kept in sync. If you change the value here because it
+    # breaks some tests, you'll definitely want to change it in
+    # basic-build-test.sh as well.
+    RELEASE_SEED=1
+
+    : ${MBEDTLS_TEST_OUTCOME_FILE=}
+    : ${MBEDTLS_TEST_PLATFORM="$(uname -s | tr -c \\n0-9A-Za-z _)-$(uname -m | tr -c \\n0-9A-Za-z _)"}
+    export MBEDTLS_TEST_OUTCOME_FILE
+    export MBEDTLS_TEST_PLATFORM
+
+    # Default commands, can be overridden by the environment
+    : ${OPENSSL:="openssl"}
+    : ${OPENSSL_LEGACY:="$OPENSSL"}
+    : ${OPENSSL_NEXT:="$OPENSSL"}
+    : ${GNUTLS_CLI:="gnutls-cli"}
+    : ${GNUTLS_SERV:="gnutls-serv"}
+    : ${GNUTLS_LEGACY_CLI:="$GNUTLS_CLI"}
+    : ${GNUTLS_LEGACY_SERV:="$GNUTLS_SERV"}
+    : ${OUT_OF_SOURCE_DIR:=./mbedtls_out_of_source_build}
+    : ${ARMC5_BIN_DIR:=/usr/bin}
+    : ${ARMC6_BIN_DIR:=/usr/bin}
+    : ${ARM_NONE_EABI_GCC_PREFIX:=arm-none-eabi-}
+    : ${ARM_LINUX_GNUEABI_GCC_PREFIX:=arm-linux-gnueabi-}
+    LIBFAKETIME=
+
+    # if MAKEFLAGS is not set add the -j option to speed up invocations of make
+    if [ -z "${MAKEFLAGS+set}" ]; then
+        export MAKEFLAGS="-j$(all_sh_nproc)"
+    fi
+
+    # Include more verbose output for failing tests run by CMake or make
+    export CTEST_OUTPUT_ON_FAILURE=1
+
+    # CFLAGS and LDFLAGS for Asan builds that don't use CMake
+    # default to -O2, use -Ox _after_ this if you want another level
+    ASAN_CFLAGS='-O2 -Werror -fsanitize=address,undefined -fno-sanitize-recover=all'
```

Comment:
Note that I'm not sure whether we _want_ to do that: it makes sense to run the faketime job without Asan, since we only care about functional testing there. But it would be convenient. to be able to at least occasionally run a full` all.sh` with faketime, so I intend to make a patch to our build/test scripts (to support faketime invocation when running tests) and to the dockerfiles (to build our own faketime instead of using the one from the Linux distribution) that makes it possible.

Reply:
I've been experimenting with that and I have dockerfile patches ready in https://github.com/Mbed-TLS/mbedtls-test/tree/dev/gilles-peskine-arm/faketime-20240116 (this branch also contains groovy patches that were just for this one test run). I also have some build script patches in https://github.com/gilles-peskine-arm/mbedtls/tree/faketime-testing-3.4.1 etc which I find too invasive. CI results: [development](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/43/), [3.4.1](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/42/), [2.28.4](https://mbedtls.trustedfirmware.org/job/mbedtls-release-ci-testing/41/).

Reply:
`ASAN_OPTIONS=verify_asan_link_order=0 faketime -f +100d  ` might be better solution. See https://github.com/google/sanitizers/wiki/AddressSanitizerFlags. 

I just pass `test_default_cmake_gcc_asan`, I test it with `LD_PRELOAD`. And I like your patches. With ASAN_OPTIONS, no invasion needed.

Reply:
In your patches, change https://github.com/Mbed-TLS/mbedtls-test/blob/5e1d2ee0d511705178d0db99b54107ed3887300a/vars/gen_jobs.groovy#L550 to 
`common.all_sh_precommand += "ASAN_OPTIONS=verify_asan_link_order=0  faketime -f +${days_ahead}d "` . I think that's enough.